### ISSUE TITLE
VS2017: completion bug on provided members with . in them

### ISSUE THREAD

**Issue by smoothdeveloper** [Microsoft/visualfsharp#2610]
I'm facing some issues with completion and colorization in VS2017 which didn't exist in previous release.

# completion bug  of TP members

When I pick members which are to be escaped with ` `` `, it doesn't escape them when committing the completion

VS2017:

![vs2017 completion](https://cloud.githubusercontent.com/assets/87944/23903349/87995f1a-08c4-11e7-87dd-a56f4403879e.gif)

VS2015:

![vs2015 completion](https://cloud.githubusercontent.com/assets/87944/23903354/8b2d0ab4-08c4-11e7-8b8d-abe1f3c84da0.gif)

(the popup don't show in capture)

# weird completion behaviour in TP arguments

trying to complete with CTRL+Space at the `.Text` with carret anyplace after `T`:

![image](https://cloud.githubusercontent.com/assets/87944/23903431/cedd30e0-08c4-11e7-9143-e68b665928f3.png)

it completes with

![image](https://cloud.githubusercontent.com/assets/87944/23903542/256eeac0-08c5-11e7-8535-6cbf40d2b5b1.png)

This is, I gather, a hidden TP generated member

# colorization of TP arguments

The colorization seems off when a type argument is using the ` `` ` names:

VS2017:
![image](https://cloud.githubusercontent.com/assets/87944/23903219/08a47988-08c4-11e7-8b8b-90fa73cba03d.png)

VS2015:
![image](https://cloud.githubusercontent.com/assets/87944/23903233/197e46b2-08c4-11e7-9de6-d2c7a536a862.png)

Might be related to #2585

attaching a minimal environment (you'll need to adjust connection string to a SQL Server available to you)
[tpcompletion.zip](https://github.com/Microsoft/visualfsharp/files/841444/tpcompletion.zip)

**Comment by vasily-kirichenko** [comment]
I've completely forgot, but it may be the case that I fixed some of these bugs in https://github.com/Microsoft/visualfsharp/pull/2235. Maybe you could resurrect it and check?

**Comment by smoothdeveloper** [comment]
@vasily-kirichenko I see, I'll try to pull that and test it with my case, thanks!

**Comment by smoothdeveloper** [comment]
memo: @cloudRoutine mentions I should look at `CompletionProvider.fs` to troubleshoot that issue

```fsharp
    override this.GetChangeAsync(_, item, _, _) : Task<CompletionChange> =
        match item.Properties.TryGetValue IsExtensionMemberPropName with
        | true, _ -> ()  // do not add extension members to the MRU list
        | _ ->
            match item.Properties.TryGetValue FullNamePropName with
            | true, fullName ->
                match mruItems.TryGetValue fullName with
                | true, hints -> mruItems.[fullName] <- hints + 1
                | _ -> mruItems.[fullName] <- 1
            | _ -> ()
        
        let nameInCode =
            match item.Properties.TryGetValue NameInCodePropName with
            | true, x -> x
            | _ -> item.DisplayText
        
        Task.FromResult(CompletionChange.Create(new TextChange(item.Span, nameInCode)))
```

**Comment by vasily-kirichenko** [comment]
@smoothdeveloper that code just replaces completion item name with `nameInCode`, which we stored into properties earlier. You should look at how we form it here https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/vs/ServiceDeclarations.fs#L1501

**Comment by smoothdeveloper** [comment]
@cloudRoutine / @vasily-kirichenko thanks for the pointers, the first bug should be fixed in #2649 (and hopefully no regression).

The second one looks a bit esoteric, hopefully it is also happening around same place.
